ELearningPlatform/src/main/java/com/example/ELearningPlatform/service/CourseService.java

```diff
@@ -0,0 +1,92 @@
+package com.example.ELearningPlatform.service;
+
+import com.example.ELearningPlatform.dto.CourseSearchRequest;
+import com.example.ELearningPlatform.entity.Course;
+import com.example.ELearningPlatform.entity.Enrollment;
+import com.example.ELearningPlatform.entity.Review;
+import com.example.ELearningPlatform.model.EnrollmentStatus;
+import com.example.ELearningPlatform.repository.CourseRepository;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.criteria.*;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Example;
+import org.springframework.data.domain.ExampleMatcher;
+import org.springframework.data.jpa.convert.QueryByExamplePredicateBuilder;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class CourseService {
+    private final CourseRepository courseRepository;
+
+    @PersistenceContext
+    private EntityManager em;
+
+    public List<Course> findCoursesAdvanced(CourseSearchRequest request) {
```

Comment:
podziel to na mniejsze metody bo zeby to przeczytac i zrozumiec to trzeba to na raty robic